### ISSUE TITLE
Clean clipboard actions not working

### ISSUE THREAD

**Issue by Chaoscontrol** [Jack251970/Flow.Launcher.Plugin.ClipboardPlus#7]
I've tried a few times but nothing seems to actually clear the database. 

[Video here. ](https://xfiles.s-ul.eu/fmnjRt8f)

**Comment by Jack251970** [comment]
Sorry, it seems that you click "clean clipboard" action, which is used to clean the data on current clipboard.
If you want to clean all data, please use "clear records" action.

![image](https://github.com/user-attachments/assets/48962d23-46a8-4c49-845e-8325aed762e3)

**Comment by Jack251970** [comment]
I will improve desc of this action. "clean clipboard data" seems to cause misunderstanding.

**Comment by Chaoscontrol** [comment]
> Sorry, it seems that you click "clean clipboard" action, which is used to clean the data on current clipboard. If you want to clean all data, please use "clear records" action.

No, the video shows me clicking the clear database option first.

Since that doesn't work, I also try the clean clipboard (I thought that referred to Clipboard+, so yeah that could do with a better description).

But the clear database option doesn't work.

**Comment by Jack251970** [comment]
Oh, I find it. Normally, when you click "Clear records", it should expand a second context menu like this:
![image](https://github.com/user-attachments/assets/cdd544b8-1be8-4f0f-a81c-18c343033b72)

**Comment by Jack251970** [comment]
And the normal behaviour of the FL should be like this:
https://github.com/user-attachments/assets/95bdd8fd-921a-4cab-953e-be7a0d070410

**Comment by Jack251970** [comment]
Could you share me with your settings?

Have you set this option to blank?
![image](https://github.com/user-attachments/assets/e95b3c34-203e-4f56-be42-50e9ecb2a579)

I find when you click the `Clear` action, the query goes to `c `?

**Comment by Chaoscontrol** [comment]
Yeah, my "Clear keyword" was empty. 

I have now changed it to have it like you, and indeed it works. It now shows the addition options like you showed, and when clicking that one it worked fine. 

![image](https://github.com/user-attachments/assets/65fde1fb-6bc8-4eef-ac98-1566151fb958)

And yes, now when clicking the main option "Clear records" it shows the sub-options as well. 

The bug is then just when the "Clear keyword" is empty. 

-----

Since "Clear records" is a menu then, maybe it should be better for the "Clean clipboard" option to be inside it?

I suggest for main menu:
Name: Clear records
Subtitle: Choose clear options

And then inside there will be:

- Clear in list
- Clear in both list and database
- Clear unpinned records
- Clear invalid records
- Clean/Clear clipboard

I think this would make it more organised and understandable. 

Also a question. What is the difference or the use between "Clear in list" and "Clear list and database"? I mean, what is the point of the difference? In what case would I want to keep the database, but clear the list?

**Comment by Jack251970** [comment]
As for `The bug is then just when the "Clear keyword" is empty.`, I think it can cause misunderstanding and I will add a message box for notifying this later. (Just waiting for the Flow.Launcher next update with new api of message box)
So **please keep this issue open**.

**Comment by Jack251970** [comment]
For `Since "Clear records" is a menu then, maybe it should be better for the "Clean clipboard" option to be inside it?`, I think `Clean clipboard` action is used to clean **system** clipboard data, and other actions in the `Clear records` are used to clean the **FL** data.

So they should not be placed in the same class. (System and app are from two levels I think)

**Comment by Jack251970** [comment]
And `What is the difference or the use between "Clear in list" and "Clear list and database"?`: Sometimes, when you want to start a new **temporary** project, it can help you clean the original list data and focus on the data of this new project.

E.g. You want to write a python notebook codes (someone ask you), you can use `Clear in list` to clean the data in the list and disable the `Keep text` option so that this temporaray data will not be saved.
Then you can copy and paste to finish this notebook, and send it to someone.
After that, you can use `Clear in list` to clean the data in the list and enable the `Keep text` option to focus on your original work.

**Comment by Jack251970** [comment]
Thanks for your finding of this empty problem! I have never thought about this before.

**Comment by Chaoscontrol** [comment]
> And `What is the difference or the use between "Clear in list" and "Clear list and database"?`: Sometimes, when you want to start a new **temporary** project, it can help you clean the original list data and focus on the data of this new project.
> 
> E.g. You want to write a python notebook codes (someone ask you), you can use `Clear in list` to clean the data in the list and disable the `Keep text` option so that this temporaray data will not be saved. Then you can copy and paste to finish this notebook, and send it to someone. After that, you can use `Clear in list` to clean the data in the list and enable the `Keep text` option to focus on your original work.

I understand, but why would you want to clear the list but KEEP the database? 

I usually find more normal that if you're happy to clear the list, you also want to clear the database. When I do it, it's just because it's starting to load quite slow. So I think it's the db that's the problem, and that's what I want to clear.

**Comment by Jack251970** [comment]
Please try to think this from the storage level of the data.

From my view, list and database represent two storage levels of the data.

The former is for the shallow, temporary data and the latter is for the deep, permanent data.

If we want to clean the shallow level of the data, I don't think we are supposed to clear the deep level of the data. (keep data for a pretty short time like the example above)

But on the contrary, if we want to clean the deep level of the data, I think we are supposed to clear the shallow level of the data. (Clear up all data because the loading is quite slow as you said)

To the end, this is just my design philosophy and I think this action has a purpose. After all, it is an available option.

**Comment by Chaoscontrol** [comment]
Definitely I see your point. I don't think it should be removed. It's good to always have both options. 

My question was more in trying to find use cases for it. It's hard for me to find a reason for a user to want to keep the database (deep inaccessible level) but clear the list only (shallow user accessible level). 

I'm just curious of the use you or other people give to this. Maybe I can find a use for myself.

**Comment by Chaoscontrol** [comment]
Do you find it getting slow to load after a while? How often do you clear your db? I don't know if that's normal and expected, or something wrong in my case.

**Comment by Jack251970** [comment]
If you have a lot of records, it can affect the FL performance.

Last time I cleared the db is one month ago when I found the memory of FL went to 2G.

Perhaps you can change the max records to a lower number so that it can automatically clear the older records when reaching the maximum number.

**Comment by Jack251970** [comment]
And I perfer to clean the records manually because I need to check those pinned records.

**Comment by Chaoscontrol** [comment]
> If you have a lot of records, it can affect the FL performance.
> 
> Last time I cleared the db is one month ago when I found the memory of FL went to 2G.
> 
> Perhaps you can change the max records to a lower number so that it can automatically clear the older records when reaching the maximum number.

Doesn't the "Keep text" etc times do the same as the "Max records" option? Which one takes priority? I currently use 1 month, 7 days, 7 days.

**Comment by Jack251970** [comment]
Sorry, I am wrong.

`Max records` helps you keep the [max] latest records in the **list** only and it will not affect the database operation (even if you records have accessed the [max]).

` times `options keep the records in certain period, it will be triggered every time when you start the FL or reload plugin data in FL.

**Comment by Chaoscontrol** [comment]
Yeah, the new version makes all the options much clearer. Thank you very much!

The bug where "clear keyword" is empty is still happening though. [See the video. ](https://xfiles.s-ul.eu/XI3JvUke) I'm clicking the option to open the sub-menu, but it doesn't open (it just moves briefly to the top of the list).

**Comment by Jack251970** [comment]
> Yeah, the new version makes all the options much clearer. Thank you very much!
> 
> The bug where "clear keyword" is empty is still happening though. [See the video. ](https://xfiles.s-ul.eu/XI3JvUke) I'm clicking the option to open the sub-menu, but it doesn't open (it just moves briefly to the top of the list).

Please wait for FL 1.20.0 release. I need new api of this version.

**Comment by Jack251970** [comment]
@Chaoscontrol Hello, since Flow Launcher 1.20.0 appears to be delayed (there are plenty of works to do😢), I’d like to address this issue earlier.

Unfortunately, setting the clear keyword to empty shouldn't be permitted because Flow Launcher trims query strings before passing them to the plugin which makes `cb` eqauls to `cb ` (no matter how many whitespaces you have typed). So I want to make sure the clear keyword is non-empty to avoid unintended behavior.

Another possible solution is to move items in the second menu (`cb clear`) to the first menu (`cb`) if the clear keyword is empty.

Let me know if you have any ideas.

**Comment by Chaoscontrol** [comment]
I'd say don't rush it for me, really. I've been using the clear word filled now ever since you told me and there's no issue. It works fine, so it's not a bother to wait. And no-one else seems to be complaining. So I'd wait for the next API.

**Comment by Jack251970** [comment]
@Chaoscontrol I’ve added a warning and now disallow setting the clear keyword to an empty value to prevent unintended behavior. Please check the latest version!
